### ISSUE TITLE
How to finetune yolo_nas from a previous specific checkpoint

### ISSUE THREAD

**Issue by ani-mal** [Deci-AI/super-gradients#1570]
### Description 

We are trying to load a checkpoint from a previous yolo_nas training to finetune further on a different dataset, different settings, different number of classes. We are not trying to resume the training like mentioned in this issue https://github.com/Deci-AI/super-gradients/issues/1139. 

I tried providing the checkpoint absolute path and the number of classes the checkpoint has given the [implementation of the method](https://github.com/Deci-AI/super-gradients/blob/34fda6c43e9607e8fd6c39e3dca60fc1e2c512cd/src/super_gradients/training/models/model_factory.py#L189-L216) `get` from `model_factory`:

```
 def load_model(self, path=None, checkpoint_num_classes=None):
        return models.get(
            self.model_architecture,
            num_classes=len(self.model_config['dataset_params']['classes']),
            checkpoint_path=path, 
            checkpoint_num_classes=checkpoint_num_classes
        ).to(self.device)
```

However, the checkpoint when it gets resolved it is not a dictionary with `key()`, an exception is being thrown here due to that. 
https://github.com/Deci-AI/super-gradients/blob/34fda6c43e9607e8fd6c39e3dca60fc1e2c512cd/src/super_gradients/training/models/model_factory.py#L232

I further verified what `torch.load("checkpoint_path.pth")` returns, and it is returning the following format:

```.py 
import torch 
state_dict = torch.load("./final_model.pth", map_location="cpu")
```

`state_dict` is:

```.sh
YoloNAS_S(
  (backbone): NStageBackbone(
    (stem): YoloNASStem(
      (conv): QARepVGGBlock(
        (nonlinearity): ReLU(inplace=True)
        (se): Identity()
        (branch_3x3): Sequential(
          (conv): Conv2d(3, 48, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
          (bn): BatchNorm2d(48, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
        )
        (branch_1x1): Conv2d(3, 48, kernel_size=(1, 1), stride=(2, 2))
        (post_bn): BatchNorm2d(48, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
        (rbr_reparam): Conv2d(3, 48, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1))
      )
    )
    (stage1): YoloNASStage(
      (downsample): QARepVGGBlock(
        (nonlinearity): ReLU(inplace=True)
        (se): Identity()
        (branch_3x3): Sequential(
          (conv): Conv2d(48, 96, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
          (bn): BatchNorm2d(96, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
        )
        (branch_1x1): Conv2d(48, 96, kernel_size=(1, 1), stride=(2, 2))
        (post_bn): BatchNorm2d(96, eps=0.001, momentum=0.03, affine=True, track_running_stats=True)
...
      (cls_dropout_rate): Identity()
      (reg_dropout_rate): Identity()
    )
  )
)
Output is truncated.
```

Which makes sense it would fail when trying to access `keys()`, however, I am not finding any documentation or implementation on the code on how to load this type of checkpoint to yolo_nas training/finetuning. 

Thank you for your help :)

**Comment by ani-mal** [comment]
I forgot to check the closed issues. I think this might be the solution I am looking for 🤔 : https://github.com/Deci-AI/super-gradients/issues/1450#issuecomment-1709637055

Let me test it...

**Comment by BloodAxe** [comment]
It looks like the whole model is picked inside the checkpoint, and not the model's state_dict.
This is quite unusual way of saving a model and is not supported. The state_dict should contain the weights of the each layer, and not the complete model.
Can you please clarify how did you save that checkpoint?

**Comment by ani-mal** [comment]
@BloodAxe we are saving the model using `mlflow` since this artifacts the model on our MLOps pipeline. We are not running the model locally:

`mlflow.pytorch.log_model(self.model)`

How would you recommend us to save the checkpoint?

**Comment by ani-mal** [comment]
To be more specific, 

```.py 
self.model =  models.get(
            self.model_architecture,
            num_classes=len(self.model_config['dataset_params']['classes']),
            checkpoint_path=path, 
            checkpoint_num_classes=checkpoint_num_classes
        ).to(self.device)
 ```
 
 so we are saving the model that is returned by the `models.get(...)` from `super_gradients.training.models`

**Comment by BloodAxe** [comment]
I see. Well, if you are saving a whole model in MLFlow than perhaps you can simply do `torch.load("./final_model.pth", map_location="cpu")` and skip `models.get` completely since you already deserialize the whole model.
This is neither tested approach nor recommended approach of storing the model, but it may work.

The canonical way is to store state dict of the model: 
```
state_dict = model.state_dict()
checkpoint = { "net": state_dict }  <- SG looks for "net" key in checkpoint 
torch.save(checkpoint, "checkpoint.pth)
```

Or alternatively you can upload a best checkpoint from the experiment directory directly (ckpt_best.pth)

**Comment by ani-mal** [comment]
@BloodAxe thank you so much, I will follow up on your suggestions and see what works. Thank you for explaining and helping me  understand what was going wrong :)

**Comment by BloodAxe** [comment]
Unfortunately we don't have out-of-the-box implementation of sg_logger for MLFlow, which would simplify the usage of SG with MLFlow. But we are open for contributions :) Let us know if it's something you may be interested in contributing ;)

**Comment by ani-mal** [comment]
The approach of bypassing the `models.get(..)` works as long as the model had the same number of classes of the future training. The approach to store a state dict works better, since I need to then leverage the `model.get(..)` to specify the new number of classes. thanks again :)

**Comment by ani-mal** [comment]
@BloodAxe I am interested in contributing to a mlflow sg logger. Do you have any documentation about contributing to this repo, and some overview/thought's about such implementation?

**Comment by BloodAxe** [comment]
@ani-mal thanks for your willingness to contribute!

To start, I suggest you reading the contribution guidelines - https://github.com/Deci-AI/super-gradients/blob/master/CONTRIBUTING.md. Most important steps are setting up signed commits and following code formatting rules (Easiest way is to setup pre-commit hooks)

Then, you can check the implementation of existing loggers: https://github.com/Deci-AI/super-gradients/blob/3e1019f4e32a47cfa16b789478dfae8bb5812418/src/super_gradients/common/sg_loggers and implement a new logger similarly.

More specificaly a new class should subclass from BaseSGLogger and override methods to store results locally and to remote (mlflow). You can check how it is done in ClearMLSGLogger or WandBSGLogger.

Once you have some draft (even if it's not finished) if can open a draft PR where we can continue discussion to help you with the rest.
